byterun/array.c

```diff
@@ -100,10 +103,13 @@ CAMLprim value caml_array_unsafe_get_float(value array, value index)
 
 CAMLprim value caml_array_unsafe_get(value array, value index)
 {
+  CAMLparam2(array, index);
+  CAMLlocal1(x);
   if (Tag_val(array) == Double_array_tag)
-    return caml_array_unsafe_get_float(array, index);
+    x = caml_array_unsafe_get_float(array, index);
   else
-    return Field(array, Long_val(index));
+    caml_read_field(array, Long_val(index), &x);
```

Comment:
Same comments as `caml_array_get_addr`. It should be worthwhile optimising for the array access.